Optional onIdle(count) method added

treblereel opened the pull request:
In some cases it's very useful to be able to get a change set after some predefined period, for example, when we use APT and we want to get a list of generated files and we
 know that the whole generation stage usually takes 500ms. In this PR i introduce the simple mechanism that calls an onIdle method  after timeout expired and passes the number of events within 
that timeout.

@mdproctor ^^

mdproctor commented on core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:
```diff
@@ -224,15 +242,28 @@ public DirectoryWatcher(
       }
       for (WatchEvent<?> event : key.pollEvents()) {
+        eventCount.incrementAndGet();
+        if (currentOnIdleTimerTask != null && !currentOnIdleTimerTask.isDone) {
```
Does this need to be in the 'for' loop? Isn't it enough to do this  directly after the take() returns?

I don't think this is thread safe, as the timer will set isDone, which is different to the directory watcher main loop thread.

mdproctor replied:
This isDone() done call is not thread safe.

mdproctor commented on core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:
```diff
@@ -224,15 +242,28 @@ public DirectoryWatcher(
+        key = watchService.poll();
+        if (key == null) {
+          if (timeout > 0) {
+            currentOnIdleTimerTask = new OnIdleTimerTask(listener, eventCount.get());
```
This creates a new OnIdleTimerTask each time. What is an OnIdleTimerTask already exists, does it need to be cancelled? Is all access got this field, and new instances to this field thread safe?

mdproctor commented on core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:
```diff
@@ -166,6 +178,10 @@ public static Builder builder() {
   private FileHasher fileHasher;
+  private int timeout = -1;
+
+  private Timer timer = new Timer("DirectoryWatcherTimer");
```
Make this final? Then it can be synchronised on.

mdproctor commented on core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:
```diff
@@ -428,4 +459,25 @@ private void notifyCreateEvent(boolean isDirectory, Path path, int count, Path r
+
+  private class OnIdleTimerTask extends TimerTask {
+
+    private boolean isDone = false;
```
There is unthread safe access to this.

mdproctor commented on core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:
```diff
@@ -428,4 +459,25 @@ private void notifyCreateEvent(boolean isDirectory, Path path, int count, Path r
+    @Override
+    public void run() {
+      logger.debug("Task performed on: " + new Date() + " with event count " + counter);
+      listener.onIdle(counter);
```
We should document that this operation is not atomic - i.e. the check on isDone() and cancel() is not atomic and may overlap with this sections onIdle() call and setting isDone. And thus it is possible that actually the timer is cancelled, while this event is being fired. I would imagine for the type of use cases this is being applied to, that is not a problem, but either way should be in in the JavaDocs.

mdproctor commented on core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:
```diff
@@ -166,6 +178,10 @@ public static Builder builder() {
   private FileHasher fileHasher;
+  private int timeout = -1;
+
+  private Timer timer = new Timer("DirectoryWatcherTimer");
```
This can be final. Further set it in the constructor. As this way you can set it null if timeout is still <= 0.

treblereel replied:
other final or constructor case, right ?

mdproctor commented on core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:
```diff
@@ -224,15 +242,28 @@ public DirectoryWatcher(
       for (WatchEvent<?> event : key.pollEvents()) {
+        eventCount.incrementAndGet();
+        if (currentOnIdleTimerTask != null && !currentOnIdleTimerTask.isDone) {
+          currentOnIdleTimerTask.cancel();
```
Should this be nulled after it's cancelled?

mdproctor commented on core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:
```diff
@@ -224,15 +242,28 @@ public DirectoryWatcher(
+        key = watchService.poll();
+        if (key == null) {
+          if (timeout > 0) {
+            currentOnIdleTimerTask = new OnIdleTimerTask(listener, eventCount.get());
```
What if this instance already exists from the previous loop iteration? Maybe it should be cancelled and nulled first?

treblereel replied:
fixed

mdproctor commented on core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:
```diff
@@ -224,15 +242,28 @@ public DirectoryWatcher(
    * is closed.
    */
   public void watch() {
+    final AtomicInteger eventCount = new AtomicInteger(0);
```
Why is this atomic? the main watch() loop  is single threaded, no other thread reads this value. It's just a value that is passed to the event handler.

treblereel replied:
yeap, agree

gmethvin reviewed (commented):
We chatted about this before @mdproctor, but I'm a bit unsure of whether it's better to implement the timeout in `DirectoryWatcher` itself, or to do it in the listener. If we handle it in the listener, we'd have to provide both `onIdle` (when waiting for new events) and `onActive` (when new events are pending), but then `DirectoryWatcher` doesn't have to worry about the thread safety issues for the timer.

In other words, we'd have something like:

```java
key = watchService.poll();
if (key == null) {
  listener.onIdle();
  key = watchService.take();
  listener.onActive();
}
```

Then you can move all the complicated timer code into a separate `DirectoryChangeListener` implementation.

gmethvin commented:
Also I think for the idle timeout you could use `ScheduledExecutorService`:

```java
ScheduledExecutorService scheduledExecutor = Executors.newSingleThreadScheduledExecutor();
ScheduledFuture<?> idleFuture = null;

void onIdle(int count) { // called when we start waiting for new events
  idleFuture = scheduledExecutor.schedule(() -> { onIdleTimedOut(count); }, idleTimeout, idleTimeoutUnit);
}

void onActive(int count) { // called when we start getting events again
  if (idleFuture != null) {
    idleFuture.cancel(true);
    idleFuture = null;
  }
}

void onIdleTimedOut(int count); // call this when the timeout has passed after `onIdle`.
```

Also with `ScheduledExecutorService` it'd be easier to keep track of multiple timeout tasks if you want to do that.

treblereel commented:
I have switched to ScheduledExecutorService and transfered all OnIdle logic into ChangeSetListener, is it closer to what we want ?

As you can see, i have implemented additional ChangeSetListener constructor that accepts timeout and consumer, because current ChangeSetListener doesn't allow to subscribe to onIdle(int) directly.


wdyt ?

gmethvin commented on core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java:
```diff
@@ -44,6 +67,11 @@ public void onEvent(DirectoryChangeEvent event) {
           throw new IllegalStateException("OVERFLOW not yet handled");
       }
+
+      if (timeout > 0 && currentTask != null) {
```
I don't think we need `timeout > 0` since all we care about is there's a task to cancel.

treblereel replied:
currentTask is null if we don't use onIdle

gmethvin commented on core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java:
```diff
@@ -13,7 +18,25 @@
-  public ChangeSetListener() {}
+  private final int timeout;
+
+  private ScheduledExecutorService service;
```
nit: can be final

treblereel replied:
ScheduledExecutorService is setted only if we need OnIdle

gmethvin commented on core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java:
```diff
@@ -13,7 +18,25 @@
+    this.timeout = timeout;
+    this.onIdleListener = onIdleListener;
+
+    if (timeout > 0) {
```
I actually think there's a valid use case for `timeout == 0`, in case you want to be notified as soon as we hit an idle state.

treblereel replied:
yeap, agree

gmethvin reviewed (commented):
Is there a reason this needs to be in `ChangeSetListener` in particular? Do you plan to use it along with the other `ChangeSetListener` functionality?

Based on your test it seems to be entirely separate functionality from the rest of `ChangeSetListener`.

treblereel commented:
@gmethvin if i am not mistaken, Mark mentioned that this functionality can be moved from DirectoryWatcher to ChangeSetListener. We are planning to use OnIdle functionality with ChangeSetListener, but, to be honest, i can't say, i am happy with my current implementation, adding listener to listener doesn't look nature.

So if you have any ideas how to improve it i ll be happy to implement it

gmethvin commented:
@mdproctor @treblereel I'm just trying to understand the overall use case so we can design the best API. This is why I'm asking how it's meant to be used. As it stands right now, it appears the idle timeout functionality could be a separate listener type rather than part of `ChangeSetListener`.

Is the idea that you want `ChangeSetListener` to retrieve a new change set _the next time the watcher is idle_? In that case, for example, you could have an API that blocks until idle and retrieves the change set of all changes since the last time the watcher was idle. On its own it just seems awkward to have the idle listener as part of `ChangeSetListener` since it doesn't seem directly related to rest of the functionality.

mdproctor commented:
I think you are right. While the onIdle notification needs to be in core. We can move the timeout notification to it's own listener. We can then ask users to add the the pair of listeners (IdleTimeOutListener, ChangeSetListener), via  composite holder. Keeping the timer logic in it's own class like that, will mean easier to maintain and test code.

mdproctor commented on core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:
```diff
@@ -0,0 +1,63 @@
+    }
+  }
+
+  public abstract void onTimeout(int count);
```
do people prefer to extend this class, or pass in a lambda? A lambda would feel more modern, less lines of code.

gmethvin replied:
I would prefer having the user pass a lambda since it allows for a cleaner API.

gmethvin commented on core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:
```diff
@@ -0,0 +1,63 @@
+
+  @Override
+  public void onIdle(int count) {
+    synchronized (lock) {
```
what do you think about making this an `AtomicReference`?

```scala
if (timeout >= 0) {
  currentTaskRef.getAndUpdate(oldTask -> {
    if (oldTask != null) {
      oldTask.cancel(false);
    }
    return service.schedule(() -> onTimeout(count), timeout, TimeUnit.MILLISECONDS);
  });
}
```

Then above you could do something like:

```java
ScheduledFuture taskToCancel = currentTaskRef.getAndSet(null);
if (taskToCancel != null) {
  taskToCancel.cancel(false);
}
```

gmethvin commented on core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:
```diff
@@ -0,0 +1,63 @@
+  private ScheduledExecutorService service;
+  private ScheduledFuture currentTask;
+
+  public OnTimeoutListener(int timeout) {
```
just for clarity I'd name it with the units: `timeoutMillis`. We could also have a length and a `TimeUnit` but not sure if that's overkill.

gmethvin replied:
Also maybe we should just throw an `IllegalArgumentException` if the timeout is negative? Is there a use case for using this listener without a timeout?

treblereel replied:
@gmethvin nope, i am pretty sure nope

gmethvin reviewed (commented):
A couple questions that came up in my review:
 1. I noticed we cancel the idle state when `onEvent` is called, but technically the watcher is no longer "idle" as soon as it starts processing a new event. There could be a delay between when the watcher receives the event and when `onEvent` fires, especially if we have hashing enabled. Does that difference matter? The solution I think would be to add a call like `onActive` at the beginning of the loop.
 2. I'm assuming most users would want to track both the idle state and the individual events, so we probably want a way to do both. One possible solution would be to have the `OnTimeoutListener` forward to another user-defined `DirectoryChangeListener`. Or we could have a way to create a composite listener from a list of listeners.

gmethvin commented:
@mdproctor I think either way there's a use case for a helper like `DirectoryChangeListener.fromListeners(listeners)`.

In this case though I was wondering if we should have something like `DirectoryChangeListener.withIdleDelay(timeoutMillis, listener)`, which would just delay the call to `listener.onIdle` with your timeout code. Not totally set on that API though.

gmethvin commented on core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  public OnTimeoutListener onTimeout(Consumer<Integer> consumer) {
+    this.consumer = consumer;
```
I'd prefer to keep the API immutable if possible, and as-is this API doesn't make it obvious that this will replace the existing consumer.

I would suggest we pass a single `Consumer onIdleTimeout` in the constructor, or in a builder.

gmethvin commented on core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+public class OnTimeoutListener implements DirectoryChangeListener {
```
Can this be final?

gmethvin commented on core/src/test/java/io/methvin/watcher/DirectoryWatcherOnIdleTest.java:
```diff
@@ -0,0 +1,214 @@
+    watcher.close();
+  }
+
+  private class CompositeListener implements DirectoryChangeListener {
```
I thought we were talking about adding a helper that creates a listener out of a list of listeners? Seems like a natural addition to the API.

gmethvin commented:
@mdproctor @treblereel I merged this along with `05c155e02871a7b6f6c73a11d58b001fd9f38ed7` which addresses my comments. Let me know if that works.
